dict/dict0dict.cc

```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
+}
+
+bool dict_table_lru::contains(std::string_view name) {
+  ib_logger(ib_stream, "contains check table: %s\n", std::string{name}.c_str());
+  return m_name_lookup.contains(name);
+}
+
+bool dict_table_lru::add(std::string_view name, std::uint64_t id, dict_table_t *table) {
+  ut_a(table);
+  ib_logger(ib_stream, "adding table: %lu, %s\n", id, std::string{name}.c_str());
+
+  auto name_itr = m_name_lookup.find(name);
+  auto id_itr = m_id_lookup.find(id);
+
+  auto found_in_name_map = name_itr != m_name_lookup.end();
+  auto found_in_id_map = id_itr != m_id_lookup.end();
+  ut_a((found_in_id_map && found_in_id_map) || (!found_in_id_map && !found_in_name_map));
+
+  if (found_in_id_map) {
+    return false;
+  }
+
+  m_size++;
+
+  if (m_size > m_capacity) {
+    evict();
+  }
+
+  m_tables.emplace_front(table);
+
+  auto cached_entry = std::make_shared<cache_entry>(cache_entry{
+    .list_itr = m_tables.begin(),
+    .name = std::string{table->name},
+    .id = table->id,
+    .evicted = false,
+  });
+
+  auto name_insertion_result = m_name_lookup.emplace(cached_entry->name, cached_entry);
+  ut_a(name_insertion_result.second);
+
+  auto id_insertion_result = m_id_lookup.emplace(id, cached_entry);
+  ut_a(id_insertion_result.second);
+
+  return true;
+}
+
+bool dict_table_lru::erase(std::string_view name) {
+  if (auto itr = m_name_lookup.find(name); itr != m_name_lookup.end()) {
+    erase(itr->second);
+    m_size--;
+  }
+
+  return false;
+}
+
+bool dict_table_lru::erase(std::uint64_t id) {
+  if (auto itr = m_id_lookup.find(id); itr != m_id_lookup.end()) {
+    erase(itr->second);
+    m_size--;
+  }
+
+  return true;
+}
+
+void dict_table_lru::releaseAll(std::function<void(dict_table_t *)> deleter) {
+  for (const auto &itr : m_id_lookup) {
+    if (itr.second->evicted) {
+      // nothing to do
+    } else {
+      deleter(*(itr.second->list_itr));
+    }
+  }
+}
+
+std::int32_t dict_table_lru::size() const {
+  return m_size;
+}
+
+dict_table_lru::stats dict_table_lru::get_stats() {
+  stats result{};
+  for (const auto &itr : m_id_lookup) {
+    if (itr.second->evicted) {
+      result.total_in_evicted_state++;
+    } else {
+      result.total_in_memory++;
+    }
+  }
+
+  return result;
+}
+
+void dict_table_lru::move_to_front(const std::shared_ptr<cache_entry> &entry) {
+  list_itr_t itr{m_tables.end()};
+  if (entry->list_itr != m_tables.end()) {
+    ut_a(!entry->evicted);
+
+    auto table = *entry->list_itr;
+    m_tables.erase(entry->list_itr);
+    m_tables.emplace_front(table);
+
+    itr = m_tables.begin();
+  } else {
+    ut_a(entry->list_itr == m_tables.end());
+    ut_a(entry->evicted);
+
+    // table was evicted, load it first
+    auto table = m_table_fetcher(entry->id);
+    ut_a(table);
+
+    m_tables.emplace_front(table);
+    itr = m_tables.begin();
+  }
+
+  ut_a(itr != m_tables.end());
+
+  entry->evicted = false;
+  entry->list_itr = itr;
+}
+
+void dict_table_lru::evict() {
```

Comment:
There are at least two problems:

1. Foreign key dependencies, you can't remove just one of the tables.
2. Some state is not persistent e.g., auto-increment. We need to preserve that even if we evict the table.

Write a unit-test to check for the above two.

Reply:
oh ya. Just realized that the state will not persistent correctly. It must be recoverable but not consistent with file block. I will play with this more.